bluepysnap/nodes.py

```diff
@@ -32,12 +32,201 @@
+                    nodes selectected by the node set.
+                - ``mapping``: Returns a CircuitNodeIds object containing nodes matching a
+                    properties filter.
+                - ``None``: return all node IDs of the circuit in a CircuitNodeIds object.
+
+        Returns:
+            CircuitNodeIds: returns a CircuitNodeIds containing all the node IDs and the
+                corresponding populations. All the explicitly requested IDs must be present inside
+                the circuit.
+
+        Raises:
+            BluepySnapError: if a population is explicitly required by the user (via CircuitNodeIds)
+            BluepySnapError: if one or more ids are explicilty asked by the user using int,
+                sequence, or CircuitNodeIds.
+
+        Examples:
+            The available group parameter values (example with 2 nodes populations pop1 and pop2):
+            >>> nodes.ids(group=None)  #  returns all CircuitNodeIds from the circuit
+            >>> node_ids = CircuitNodeIds.create_ids(["pop1", "pop2"], [1, 3])
+            >>> nodes.ids(group=node_ids)  #  returns ID 1 from pop1 and ID 3 from pop2
+            >>> nodes.ids(group=0)  #  returns CircuitNodeIds 0 from pop1 and pop2
+            >>> nodes.ids(group=[0, 1])  #  returns CircuitNodeIds 0 and 1 from pop1 and pop2
+            >>> nodes.ids(group="node_set_name")  # returns CircuitNodeIds matching node set
+            >>> nodes.ids(group={Node.LAYER: 2})  # returns CircuitNodeIds matching layer==2
+            >>> nodes.ids(group={Node.LAYER: [2, 3]})  # returns CircuitNodeIds with layer in [2,3]
+            >>> nodes.ids(group={Node.X: (0, 1)})  # returns CircuitNodeIds with 0 < x < 1
+            >>> # returns CircuitNodeIds matching one of the queries inside the 'or' list
+            >>> nodes.ids(group={'$or': [{ Node.LAYER: [2, 3]},
+            >>>                          { Node.X: (0, 1), Node.MTYPE: 'L1_SLAC' }]})
+            >>> # returns CircuitNodeIds matching all the queries inside the 'and' list
+            >>> nodes.ids(group={'$and': [{ Node.LAYER: [2, 3]},
+            >>>                           { Node.X: (0, 1), Node.MTYPE: 'L1_SLAC' }]})
+        """
+        str_type = "<U{}".format(max(len(pop) for pop in self.population_names))
+        if isinstance(group, CircuitNodeIds):
+            diff = np.setdiff1d(group.get_populations(unique=True), self.population_names)
+            if diff.size != 0:
+                raise BluepySnapError("Population {} does not exist in the circuit.".format(diff))
+        ids = np.array([], dtype=np.int64)
+        populations = np.array([], dtype=str_type)
+        for name, pop in self.items():
+            pop_ids = pop.ids(group=group, raise_missing_property=False)
+            pops = np.array(np.full_like(pop_ids, fill_value=name, dtype=str_type))
+            ids = pop_ids if ids.size == 0 else np.concatenate([ids, pop_ids])
```

Comment:
Does it make sense to have this `if ids.size == 0`? It should work without it. The code will be easier to read. The same in the line below with `populations` .

Reply:
Indeed you are right. Actually I remember I had problem with this before I added the  `raise_missing_property=False` due to incorrectly initialized pop_ids. 
I just did not rethink the whole function after adding the flag. 

I removed the if else this is much better.